encoding/gzip/gzip.go

```diff
@@ -0,0 +1,79 @@
+/*
+ *
+ * Copyright 2017 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package gzip
+
+import (
+	"compress/gzip"
+	"io"
+	"io/ioutil"
+	"sync"
+
+	"google.golang.org/grpc/encoding"
+)
+
+// New function registers gzip compressor. It maintains two pools for compressor and decompressor.
+func New() encoding.Compressor {
+	return &compressor{
+		name: "gzip_register",
+		poolCompressor: sync.Pool{
+			New: func() interface{} {
+				return gzip.NewWriter(ioutil.Discard)
+			},
+		},
+		poolDecompressor: sync.Pool{},
+	}
+}
+
+func (c *compressor) Compress(w io.Writer) (io.WriteCloser, error) {
+	z := c.poolCompressor.Get().(*gzip.Writer)
+	z.Reset(w)
+	return z, nil
+}
+
+func (c *compressor) Decompress(r io.Reader) (io.Reader, error) {
+	var z *gzip.Reader
+	switch maybeZ := c.poolDecompressor.Get().(type) {
+	case nil:
+		newZ, err := gzip.NewReader(r)
+		if err != nil {
+			return nil, err
+		}
+		z = newZ
+	case *gzip.Reader:
+		z = maybeZ
+		if err := z.Reset(r); err != nil {
+			c.poolDecompressor.Put(z)
+			return nil, err
+		}
+	}
+	defer func() {
+		c.poolDecompressor.Put(z)
```

Comment:
This will re-add the decompressor to the pool before it is even returned to the caller.

We need to override `Read()` instead, so that when the underlying reader returns `io.EOF`, we do `poolDecompressor.Put()` at that time.